### PR TITLE
add to retain self

### PR DESCRIPTION
**Pull request by tony7126**
*Issue #225 :*

*Description of changes:*
Add withExtendedLifetime to retain self until potential delegate operation has been executed

### CODE REVIEW

File: AWSAppSyncClient/Internal/AWSPerformOfflineMutationOperation.swift [hunk 1]
```diff
@@ -64,22 +64,23 @@ final class AWSPerformOfflineMutationOperation: AsynchronousOperation, Cancellab
 
     private func notifyCompletion(_ result: JSONObject?, error: Error?) {
         operationCompletionBlock?(self, error)
+        withExtendedLifetime(self) {
+            handlerQueue.async { [weak self] in
```
Comment (sunchunqiang) [inline AWSAppSyncClient/Internal/AWSPerformOfflineMutationOperation.swift]: Can you explain how the change works? 
withExtendedLifetime will make sure that self will not destroyed before its closure return, but here handlerQueue.async  will return before handlerQueue.async's closure return, so withExtendedLifetime still cannot make sure the closure of handlerQueue.async  is finished before self is destroyed

Comment (tony7126) [reply AWSAppSyncClient/Internal/AWSPerformOfflineMutationOperation.swift]: I was under the impression that the scope (self) would be maintained as long as anything was unfinished with "withExtendedLifetime".  Is that wrong?  (If so can you link me to an explanation of this, if possible?).

Maybe a simpler option would be to add a third argument to notifyCompletion, a callback function.  Then the state of the object can be switched to finished in there.  I can make the update and push today if you think that would make more sense

Comment (tony7126) [reply AWSAppSyncClient/Internal/AWSPerformOfflineMutationOperation.swift]: Ok I did a toy example just to confirm you were right (which you are):
https://gist.github.com/tony7126/16de12ea7360ebda9344d3ec413d9ba5

Comment (tony7126) [reply AWSAppSyncClient/Internal/AWSPerformOfflineMutationOperation.swift]: Here is a working toy example of the callback solution:
https://gist.github.com/tony7126/7160067f7d7bc48cedf68c4421744f1d

Let me know if you think something similar to that makes sense

**Comment by palpatim** [comment]
Assigning this to @rohandubal since this may conflict with recent mutation queue updates that they did.

**Comment by rohandubal** [comment]
Hello @tony7126 

I had a look at the code changes and the gist you posted(nicely detailed and super helpful), I agree with code changes in the PR. What I still want to discuss is the primary objective of making this change. Why do you think it is beneficial to wait till the delegate notification is completely done before updating the status? I think that has a chance of opening up to certain race condition when `cancel` is invoked by the developer? What are your thoughts?

We would also need to rebase this branch if it has to be pulled in, but that can wait till we have the discussion about the purpose.

Best,
Rohan

**Comment by tony7126** [comment]
Hey Rohan,
I assume you're referring to this check on the isCancelled flag (I only see 2 and the other looks pretty clearly safe)?
https://github.com/awslabs/aws-mobile-appsync-sdk-ios/blob/13c6a5026980a70dc50881e6f35a4314fffc0fc7/AWSAppSyncClient/Internal/AWSPerformOfflineMutationOperation.swift#L109

If the error is nil it should never hit that isCancelled check and there shouldn't be a race?  Or is there another possibility I'm missing?

**Comment by rohandubal** [comment]
I agree with your changes and seems the right thing to do. Could you please rebase on top of current master?

**Comment by tony7126** [comment]
rebased.  let me know if everything looks good.

**Comment by tony7126** [comment]
@rohandubal 
any word here?

**Comment by tony7126** [comment]
@rohandubal 
Just following up again.  Any reason this can't be merged in?

**Comment by tony7126** [comment]
@rohandubal 
Not sure what should be done here.  Do you have any feedback?

**Comment by tony7126** [comment]
Hi, I've looked at the latest master and it seems this logic still isn't in there.  Is there a reason not to add this in?  This addresses something I've definitely seen happen.

**Comment by tony7126** [comment]
@rohandubal is there any reason to keep this open?

**Pull request closed without merging**